postconfirmd.1
." Text automatically generated by txt2man-1.4.7
.TH untitled  "July 10, 2017" "" ""

.SH NAME
\fBpostconfirmd.py \fP- Mail confirmation system (daemon part)
\fB
.SH SYNOPSIS
.nf
.fam C
postconfirmd.py [\fIOPTIONS\fP]
.fam T
.fi
.SH DESCRIPTION
postconfirmd.py is the long-running (daemon) part of a client-server
program which handles email confirmations. It is intended as a
front-end to mailing lists. It provides funcitonality which is a
subset of TMDA, but is adapted to high-volume usage and does not have
anywhere near all the bells and whistles which TMDA has. On the other
hand, since the allowlist lookup is done by the long-running server
part, the overhead of doing a verification that a poster has a
confirmed address is much smaller than for TMDA.
.PP
Early tests indicate that with a allowlist of 16000 entries a lookup
costs about 0.01 seconds and about ~5 Mbytes for the server daemon, in
contrast whith TMDA which cost between ~10s and ~600s and ~98 Mb *per
lookup* as it was set up in front of the IETF lists.
.PP
Somewhat simplified, the flow for regular mail is as follows:
.PP
.nf
.fam C
           mail agent
               |
               v
           postconfirm/mailman wrapper [copy] -> cached in cache directory
               |
               v
           postconfirmc <==> postconfirmd
               |
               v
           postconfirm/mailman wrapper [allow] -> Mailman
             [grey]
               |
               v
           (waiting for confirmation)

.fam T
.fi
In the start of the flow above, a message comes in. If the envelope
From is in the lists known by postconfirmd, it is passed through the
mailman/ postconfirm wrapper, where postconfirmc is called to talk
with postconfirmd and get either an OK to pass the message to Mailman,
or cache it.
.PP
In more detail, this is the sequence of actions that postconfirm goes
through when processing an incoming email message:
.IP 1. 4
The mail is put in the cache directory.
.IP 2. 4
If the message is a delivery-status report (content-type
multipart/report, report-type delivery-status) then this is logged,
and no further action is taken.
.IP 3. 4
If the precedence matches the bulk_regex setting in the
configuration file, then:
.RS
.PP
a. if the envelope sender is in the allowlist or the allow_regex
list, then the message is forwarded directly.
.PP
b. if not, the receipt of a bulk message is logged, and no further
action is taken.
.RE
.IP 4. 4
If the message has an auto-submitted header, and the value matches
the auto_submitted_regex in the configuration file, then:
.RS
.PP
a. if the envelope sender is in the allowlist or the allow_regex
list, then the message is forwarded directly.
.PP
b. if not, the receipt of a bulk message is logged, and no further
action is taken.
.RE
.IP 5. 4
If the subject matches the regex for postconfirm's confirmation
email subject, the confirmation is processed, and then:
.RS
.PP
a. if the confirmation is valid, the held message is forwarded,
and the confirmed address is added to the allowlist.
.PP
b. if not, the failed confirmation is logged, and the message
is saved.
.RE
.IP 6. 4
If the envelope sender is in the allowlist or the allow_regex list,
the message is forwarded.
.IP 7. 4
If there was no match earlier, a confirmation request is sent out
for the message.
.RE
.PP

.SH OPTIONS
.TP
.B
\fB-d\fP, \fB--debug\fP
Run with debug output
.TP
.B
\fB-f\fP, \fB--foreground\fP
Run in the foreground, don't daemonize
.TP
.B
\fB-h\fP, \fB--help\fP
Output this help, then exit
.TP
.B
\fB-s\fP, \fB--socket\fP
Specify the socket path
.TP
.B
\fB-V\fP, \fB--version\fP
Output version, then exit
.SH FILES
postconfirmd.py reads its configuration from the following files (if found),
in the given order:
.PP
.nf
.fam C
            <bin-dir>/postconfirm.conf
            /etc/postconfirm.conf
            /etc/postconfirm/postconfirm.conf
            ~/.postconfirmrc

.fam T
.fi
where <bin-dir> is the directory where postconfirmd is installed.
.SH AUTHOR
Written by Henrik Levkowetz, <henrik@merlot.tools.ietf.org>. Uses the
daemonize module from Chad J. Schroeder, from the Python Cookbook at
http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/278731, and
the readyexec module from Frank J. Tobin, available at
http://readyexec.sourceforge.net/.
.SH COPYRIGHT
Copyright 2008 Henrik Levkowetz
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version. There is NO WARRANTY; not even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE. See the GNU General Public License for more details.
.RE
.PP



changelog
postconfirm (0.42.1) ietf; urgency=low

  * Added domain name normalization (to lowercase) before checking whether we
    got a hit on the _dmarc domain lookup.  Commented out verbose wall-time
    log messages.

 -- Henrik Levkowetz <henrik@levkowetz.com>  31 Jul 2018 08:48:14 -0700

postconfirm (0.42.0) ietf; urgency=high

  * Added a check on the sender address during dmarc-reverse processing,
    requiring that the sender is whitelisted or has a confirmed address.

  * Changed the logging of dmarc-reverse processing to use a more compact
    log line format, and commented out some debug logging statements.

  * Added exception catching and loggin for failures to open smtp sessions
    to the configured server and port.

  * Fixed a log writing bug.

 -- Henrik Levkowetz <henrik@levkowetz.com>  09 May 2018 14:04:31 -0700

postconfirm (0.41.5) ietf; urgency=high

  * Fixed dmarc_reverse() to not re-send to other addresses than the 
    indicated recipient.

 -- Henrik Levkowetz <henrik@levkowetz.com>  20 Apr 2018 05:04:44 -0700


postconfirm (0.41.4) ietf; urgency=high

  * Fixed a bug in dmarc_rewrite(), where the return values from
    email.utils.getaddresses() was not handled properly.

  * Tweaked some setup error messages

 -- Henrik Levkowetz <henrik@levkowetz.com>  01 Mar 2018 06:47:58 -0800

postconfirm (0.41.3) ietf; urgency=high

  * Added warnings on failure for some post-install actions.

  * Changed the interface to accept multiple recipients when given on the
    command line (for use with dmarc-rewrite).  Changed dmarc-rewrite to
    accept multiple recipients, and renamed the recipient argument for other
    functions that can accept recipient lists.  

  * Changed the -e argument to log arguments to syslog and not exit.

  * Reverted an earlier change of socket ownership handling.

 -- Henrik Levkowetz <henrik@levkowetz.com>  21 Feb 2018 08:29:30 -0800

postconfirm (0.41.2) ietf; urgency=high

  * Added and --echo argument to echo back sys.argv.  Rewrote the logic to 
    find sender and recipient, in order to not throw out SENDER and RECIPIENT 
    in environment when they are set to nothing, which in some circumstances 
    can be valid.

  * Don't enforce uid and gid  on socket in foreground mode, to make 
    testing easier.

  * Corrected the trove classifier for development status.

 -- Henrik Levkowetz <henrik@levkowetz.com>  16 Feb 2018 04:28:30 -0800

postconfirm (0.41.1) ietf; urgency=high

  * Changed default dmarc-rewrite %-encoding character from '%' to '=', and 
    made it configurable.

 -- Henrik Levkowetz <henrik@levkowetz.com>  15 Feb 2018 09:57:40 -0800

postconfirm (0.41.0)

  This the first release uploaded to pypi.python.org, for more standardised
  installation using pip.  It's still not expected to be ready for beta use,
  but installation on the IETF servers should be easier this way.

  The essensial feature of this release is the ability to do experimental
  dmarc rewrites, both for outgoing mailman list messages, and to handle
  replies coming in to the rewritten addresses.  From the commit log:

  * Added dmarc rewrite functionality to postconfirmd.  Also various 
    accumulated tweaks and changes.

  * Refined setup, in preparation for alpha release to pypi.

  * README updated to indicate that this is an alpha release.

  * Refined mailman interface wrapper, with conditional filter actions.

  * Added some new config defaults.  Fixed an issue with the help display.  
    Updated imports.  Added chown for the socket.

  * Refactored service.py to permit multiple actions.  Added option 
    parsing.  Added bounce handling, but commented out.  Added dmarc handling.  
    Added comprehensive docstring for --help messages.

  * Updated the docstring.  Added default config for dmarc.

  * Tweaked sockserver output to straghten multiple lines into one.

 -- Henrik Levkowetz <henrik@levkowetz.com>  14 Feb 2018 16:01:53 +0000


postconfirm (0.40)

  * Added checking of the mailman list archive flag before adding the
    Archived-At: header field.

  * Added a mailman_dir config option, in order to be able to import
    the python mailman libs even if not on the regular python path.

  * Fixed the -s / --socket option, to make it take an argument as intended.

  * Changed the handling of sys.stdout and sys.stderr when running in foreground
    mode -- they should not go to syslog in that case.

  * Added reading of mailman listinfo for all mailman lists on startup and
    on receipt of SIGHUP.

 -- Henrik Levkowetz <henrik@levkowetz.com>  20 Jan 2017 10:36:45 +0100

postconfirm (0.39)

  * Added support for the Archived-At: header field for the IETF MailArchive
    utility

 -- Henrik Levkowetz <henrik@levkowetz.com>  10 Feb 2014 21:28:27 +0100

postconfirm (0.38)

  * Added BATV (draft-levine-smtp-batv) support.

 -- Henrik Levkowetz <henrik@levkowetz.com>  20 Mar 2010 16:49:50 -0700

postconfirm (0.37)

  * Tweaked the log message for dropped bulk messages

  * Added detection of auto-reply messages which contain a valid
    confirmation code in the body of the message, but not in the
    subject line.

 -- Henrik Levkowetz <henrik@levkowetz.com>  25 Sep 2009 15:37:17 +0200

postconfirm (0.36)

  * Moved the bulk mail detection to after the whitelist check.  Some
    OK mail is also bulk mail.

  * Added logging of py module loading, indicating which file it comes from.

 -- Henrik Levkowetz <henrik@levkowetz.com>  25 Sep 2009 11:47:24 +0200

postconfirm (0.35)

  * Tweaked the wrapper to make the 'filter' action work

  * Added detection of bulk mail and a config file option to set
    a regex for the Precedence: header to detect bulk mail.

 -- Henrik Levkowetz <henrik@levkowetz.com>  24 Sep 2009 17:07:41 +0200

postconfirm (0.33)

  * Added a new action, 'filter', which forwards confirmed email to
    /usr/lib/sendmail instead of to mailman.  For use with filtering
    of alias addresses.


postconfirm/postconfirmd.py
#!/usr/bin/python
# -*- python -*-

"""
NAME
	%(program)s - Mail confirmation system (daemon part)

SYNOPSIS
	%(program)s [OPTIONS]

DESCRIPTION
        %(program)s is the long-running (daemon) part of a client-server
        program which handles email confirmations. It is intended as a
        front-end to mailing lists. It provides funcitonality which is a
        subset of TMDA, but is adapted to high-volume usage and does not have
        anywhere near all the bells and whistles which TMDA has. On the other
        hand, since the allowlist lookup is done by the long-running server
        part, the overhead of doing a verification that a poster has a
        confirmed address is much smaller than for TMDA.

        Early tests indicate that with a allowlist of 16000 entries a lookup
        costs about 0.01 seconds and about ~5 Mbytes for the server daemon, in
        contrast whith TMDA which cost between ~10s and ~600s and ~98 Mb *per
        lookup* as it was set up in front of the IETF lists.

        Somewhat simplified, the flow for regular mail is as follows:

           mail agent
               |
               v
           postconfirm/mailman wrapper [copy] -> cached in cache directory
               |
               v
           postconfirmc <==> postconfirmd
               |
               v
           postconfirm/mailman wrapper [allow] -> Mailman
             [grey]
               |
               v
           (waiting for confirmation)

        In the start of the flow above, a message comes in. If the envelope
        From is in the lists known by postconfirmd, it is passed through the
        mailman/ postconfirm wrapper, where postconfirmc is called to talk
        with postconfirmd and get either an OK to pass the message to Mailman,
        or cache it.
        
        In more detail, this is the sequence of actions that postconfirm goes
        through when processing an incoming email message:

        1. The mail is put in the cache directory.

        2. If the message is a delivery-status report (content-type
           multipart/report, report-type delivery-status) then this is logged,
           and no further action is taken.

        3. If the precedence matches the bulk_regex setting in the
           configuration file, then:
           
           a. if the envelope sender is in the allowlist or the allow_regex
               list, then the message is forwarded directly.

           b. if not, the receipt of a bulk message is logged, and no further
               action is taken.

        4. If the message has an auto-submitted header, and the value matches
           the auto_submitted_regex in the configuration file, then:
           
           a. if the envelope sender is in the allowlist or the allow_regex
              list, then the message is forwarded directly.

           b. if not, the receipt of a bulk message is logged, and no further
              action is taken.

        5. If the subject matches the regex for postconfirm's confirmation
           email subject, the confirmation is processed, and then:
           
           a. if the confirmation is valid, the held message is forwarded,
              and the confirmed address is added to the allowlist.

           b. if not, the failed confirmation is logged, and the message
              is saved.

        6. If the envelope sender is in the allowlist or the allow_regex list,
           the message is forwarded.

        7. If there was no match earlier, a confirmation request is sent out
           for the message.


%(options)s

FILES
        %(program)s reads its configuration from the following files (if found),
        in the given order:
        
            <bin-dir>/postconfirm.conf
            /etc/postconfirm.conf
            /etc/postconfirm/postconfirm.conf
            ~/.postconfirmrc

        where <bin-dir> is the directory where postconfirmd is installed.

AUTHOR
	Written by Henrik Levkowetz, <henrik@merlot.tools.ietf.org>. Uses the
        daemonize module from Chad J. Schroeder, from the Python Cookbook at
        http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/278731, and
        the readyexec module from Frank J. Tobin, available at
        http://readyexec.sourceforge.net/.

COPYRIGHT
	Copyright 2008 Henrik Levkowetz

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or (at
	your option) any later version. There is NO WARRANTY; not even the
	implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
	PURPOSE. See the GNU General Public License for more details.

"""

import re
import sys
import os.path
import stat
import getopt
import config
import StringIO
import syslog
from __init__ import __version__

# ------------------------------------------------------------------------------
# Misc. metadata


program = os.path.basename(sys.argv[0])
progdir = os.path.dirname(sys.argv[0])

# ------------------------------------------------------------------------------
# Set up logging

syslog.openlog("postconfirmd", syslog.LOG_PID)
log = syslog.syslog

log("Postconfirm daemon v%s starting (%s)." % (__version__, __file__))

# ------------------------------------------------------------------------------
# Read config file


merger = config.ConfigMerger(lambda x, y, z: "overwrite")

default = """
foreground:         False
debug:              False
socket_path:        "/var/run/postconfirm/socket"
archive_url_pattern: "http://mailarchive.ietf.org/arch/msg/%(list)s/%(hash)s"
auto_submitted_regex:	"^auto-"
mailman_dir:        "/usr/lib/mailman"
remember_bounce_hours: 12
confirm:
{
    smtp:
    {
        host: localhost
        port: 25
    }
}
dmarc:
{
    domain:         "dmarc.ietf.org"
    resolver:
    {
        # seconds
        timeout: 3                      
        lifetime: 5
    }
    rewrite:
    {
        quote_char: "="
        smtp:
        {
            host: localhost
            port: 10028
        }
    }
    reverse:
    {
        smtp:
        {
            host: localhost
            port: 10025
        }
    }
}
"""

default = StringIO.StringIO(default)
conf = config.Config(default)

for conffile in [progdir+"/postconfirm.conf",
                "/etc/postconfirm.conf",
                "/etc/postconfirm/postconfirm.conf",
                os.path.expanduser("~/.postconfirmrc"),
            ]:
    if os.path.exists(conffile):
        log("Reading configuration from %s" % conffile)
        merger.merge(conf, config.Config(conffile))
try:
    conf.dmarc.rewrite.quote_char
except AttributeError:
    conf.dmarc.rewrite.quote_char = '='

# import pprint
# pprint.pprint(dict(conf), sys.stdout, 4, 40)

# ------------------------------------------------------------------------------
# Create list of options, for the help text

options = ""
for line in re.findall("\n +(if|elif) +opt in \[(.+)\]:\s+#(.+)\n", open(__file__.replace(".pyc", ".py")).read()):
    if not options:
        options += "OPTIONS\n"
    options += "        %-16s %s\n" % (line[1].replace('"', ''), line[2])
options = options.strip()

# ------------------------------------------------------------------------------
# Process options

# with ' < 1:' on the next line, this is a no-op:
if len(sys.argv) < 1:
    print __doc__ % locals()
    sys.exit(1)

try:
    opts, args = getopt.gnu_getopt(sys.argv[1:], "dfhs:V", ["debug", "foreground", "help", "socket=", "version", ])
except Exception, e:
    print "%s: %s" % (program, e)
    sys.exit(1)

# process option switches
for opt, value in opts:
    if not opt:
        pass
    elif opt in ["-d", "--debug"]:      # Run with debug output
        conf.debug = True
    elif opt in ["-f", "--foreground"]: # Run in the foreground, don't daemonize
        conf.foreground = True
    elif   opt in ["-h", "--help"]:     # Output this help, then exit
        print __doc__ % globals()
        sys.exit(1)
    elif opt in ["-s", "--socket"]:     # Specify the socket path
        conf.socket_path = value
    elif opt in ["-V", "--version"]:    # Output version, then exit
        print program, __version__
        sys.exit(0)


# ------------------------------------------------------------------------------

import daemonize
import grp
import pwd
import service
import sockserver
import syslog

# ------------------------------------------------------------------------------
# Utility functions

def mkdir(path):
    if not os.path.exists(path):
        print "Creating directory '%s'" % path
        os.mkdir(path)
        return True
    return False

def mkfile(file):
    path = os.path.dirname(file)
    mkdir(path)
    if not os.path.exists(file):
        os.mknod(file)
        return True
    return False

# ------------------------------------------------------------------------------
# Some assertions
assert os.path.exists(conf.mail_template), "Could not find mail template at configured path: %s" % conf.mail_template
assert type(conf.allowlists) == config.Sequence, "The allowlist configuration should be a list of file paths."

# ------------------------------------------------------------------------------
# Create directories and do initialization, as needed:

def run():
    sys.stderr.write("\nPostconfirm daemon v%s starting (%s).\n" % (__version__, __file__))

    mkdir(conf.mail_cache_dir)

    if mkfile(conf.key_file):
        keyfile = open(conf.key_file, "w")
        keyfile.write(os.urandom(128))
        keyfile.close()

    mkfile(conf.confirmlist)
    #mkfile(conf.bouncelist)

    mkdir(os.path.dirname(conf.socket_path))



    # ------------------------------------------------------------------------------
    # Maybe daemonize

    # Get user and group we should execute as
    user, pw, uid, gid, gecos, home, shell = list(pwd.getpwnam(conf.daemon_user))
    if "daemon_group" in conf:
        group, gpw, gid, members = list(grp.getgrnam(conf.daemon_group))

    if not conf.foreground:
        pidfname = "/var/run/%s.pid" % program

        pidfile = open(pidfname, "w")
        os.chown(pidfile.name, uid, gid)
        pidfile.close()

        os.setgid(gid)
        os.setuid(uid)
        daemonize.createDaemon()

        pidfile = open(pidfname, "w")
        pidfile.write("%s" % os.getpid())
        pidfile.close()
    else:
        log("Not daemonizing.")
        sys.stderr.write("Not daemonizing.\n")

    sys.stderr.write("Will listen on unix domain socket '%s'\n" % conf.socket_path)

    # ------------------------------------------------------------------------------
    # connect stdout and stderr to syslog, to catch messages, exceptions, traceback
    if not conf.foreground:
        log("Redirecting stdout and stderr to syslog")
        syslog.write = syslog.syslog
        sys.stdout = syslog
        sys.stderr = syslog

    # ------------------------------------------------------------------------------
    # Set up the service
    service.setup(conf, args)

    # ------------------------------------------------------------------------------
    # Start the server

    try:
        server = sockserver.ReadyExec(service.handler, conf.socket_path, debug=conf.debug)
    except IOError as e:
        log(" ** Tried to set up server to read from socket %s, but got an exception: '%s'" % (conf.socket_path, e))
        raise

    os.chmod(conf.socket_path, stat.S_IRWXU | stat.S_IRWXG )
    os.chown(conf.socket_path, uid, gid)

    try:        
        server.serve_forever()
    finally:
        server.server_close()

if __name__ == '__main__':
    run()
    
